### PR TITLE
Fix save carrier with invalid ranges

### PR DESCRIPTION
**Pull request by Azouz-Jribi**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When you create or edit a carrier with invalid ranges, you should not be able to save it.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-3155
| How to test?  | Go to BO > shipping > carrier > Edit , add a new range and fill its input with incorrect values. Then, try to click another input field, the incorrect fields will be red, error will be displayed and the submit button will be disabled.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8097)

### CODE REVIEW

**Comment by vincentbz** [comment]
Hi @Azouz-Jribi 

Could you move it to 1.7.2.x branch please ? 
So that it can be taken in 1.7.2.1 version.

Thanks !

File: js/admin/carrier_wizard.js [hunk 1]
```diff
@@ -403,6 +400,14 @@ function bind_inputs()
 			return false;
 	});
 
+	$('tr.range_sup td input:text, tr.range_inf td input:text').blur(function ()
+	{
```
Comment (maximebiloe) [inline js/admin/carrier_wizard.js]: Please, put the bracket on the previous line

**Review by maximebiloe — CHANGES_REQUESTED** [review]

File: js/admin/carrier_wizard.js (hunk 1 above)
Comment (Azouz-Jribi) [reply js/admin/carrier_wizard.js]: Hi @maximebiloe 
Correction is done

regards

**Review by maximebiloe — DISMISSED** [review]

**Comment by marionf** [comment]
Hello @Azouz-Jribi 
There is a problem when I try to change a range or when I click several times in the different fields.
The page freezes.

pr8097

**Comment by vincentbz** [comment]
Hi @Azouz-Jribi 

could you check this with @marionf  please ? 

Thanks !

**Comment by Azouz-Jribi** [comment]
Hi @marionf,
This bug was already exist from the beginning, but with the new modification will be more reproducible.

bug js

After some change the page crashes, see the GIF.

File: js/admin/carrier_wizard.js [hunk 2]
```diff
@@ -403,6 +400,13 @@ function bind_inputs()
 			return false;
 	});
 
+	$('tr.range_sup td input:text, tr.range_inf td input:text').blur(function () {
```
Comment (eternoendless) [inline js/admin/carrier_wizard.js]: This selector is written at least twice elsewhere (lines 410 and 569). Please consider writing an encapsulating function somewhere else:
```js
var carrier = {
  getRangeBoundaryInputs: function() {
     return $('tr.range_sup td input:text, tr.range_inf td input:text');
  }
};
```
and then
```js
var rangeBoundaryInputs = carrier.getRangeBoundaryInputs();
rangeBoundaryInputs
  .blur(...)
  .typeWatch(...);

File: js/admin/carrier_wizard.js [hunk 3]
```diff
@@ -553,15 +554,24 @@ function validateRange(index)
...
-		{
+		if (!is_valid) {
 			$('tr.range_sup td:eq('+index+')').find('div.input-group input:text').closest('div.input-group').addClass('has-error');
 			$('tr.range_inf td:eq('+index+')').find('div.input-group input:text').closest('div.input-group').addClass('has-error');
 			displayError([range_is_overlapping], $("#carrier_wizard").smartWizard('currentStep'));
-		}
-		else
+		} else {
+			$('tr.range_sup td input:text').closest('div.input-group').removeClass('has-error');
+			$('tr.range_inf td input:text').closest('div.input-group').removeClass('has-error');
```
Comment (eternoendless) [inline js/admin/carrier_wizard.js]: You can merge these two lines into one using the function I suggested.

File: js/admin/carrier_wizard.js [hunk 4]
```diff
@@ -792,3 +802,16 @@ function checkAllZones(elt)
...
+ * @param index
+ */
+function checkCarriersRange(index)
+{
+	if (validateRange(index)) {
+		enableRange(index);
+	} else {
+		disableRange(index);
+	}
+}
```
Comment (eternoendless) [inline js/admin/carrier_wizard.js]: All new methods should be "namespaced" (that is, not in the global scope).

One "namespacing" technique is to make functions be properties of a namespace variable (see the example I provided before).

**Comment by marionf** [comment]
Hello @Azouz-Jribi
I can't validate your PR, the page freezes to easily.
Also, some fields are disabled and I can't enter the price for the range.
I attached a new screenrecord on the forge ticket to show you.

**Comment by Azouz-Jribi** [comment]
Hi @marionf 
i will try to fix it, but  it takes time.

**Comment by Azouz-Jribi** [comment]
Hello @marionf 
Correction is done.
Can you retest it please ?

Thanks ;)

File: js/admin/carrier_wizard.js [hunk 5]
```diff
@@ -792,3 +779,48 @@ function checkAllZones(elt)
 	}
 
 }
+
+/**
+ * Check the carriers range inputs after each change
+ */
+const checkCarriersRangeInput = function checkCarriersRangeInput() {
```
Comment (eternoendless) [inline js/admin/carrier_wizard.js]: We cannot use `const` in legacy ES5 files, ES6 features can only be used in "babelized" files.

Comment (eternoendless) [reply js/admin/carrier_wizard.js]: The whole block for this function is wrongly intended. You should use two spaces for indents.

File: js/admin/carrier_wizard.js [hunk 6]
```diff
@@ -553,15 +533,22 @@ function validateRange(index)
...
-		{
+		if (!is_valid && $range.closest('div.input-group.has-error').length === 0) {
 			$('tr.range_sup td:eq('+index+')').find('div.input-group input:text').closest('div.input-group').addClass('has-error');
 			$('tr.range_inf td:eq('+index+')').find('div.input-group input:text').closest('div.input-group').addClass('has-error');
 			displayError([range_is_overlapping], $("#carrier_wizard").smartWizard('currentStep'));
-		}
-		else
+		} else if (is_valid && $range.closest('div.input-group.has-error').length === 0) {
 			isOverlapping();
+		}
```
Comment (eternoendless) [inline js/admin/carrier_wizard.js]: You can avoid repeating yourself:

```js
if ($range.closest('div.input-group.has-error').length === 0)) {
  if (is_valid) {
    // ...
  } else {
    // ..
  }
}

**Review by eternoendless — DISMISSED** [review]
@Azouz-Jribi QA needs to wait for code review before they can verify tickets.

**Comment by Azouz-Jribi** [comment]
Hi @eternoendless 
Correction is done, can you check it ?

Thanks

**Comment by marionf** [comment]
Hello @Azouz-Jribi
When I try to change a range, even if the range isn't overlapping anymore, the fields stay in red and there is an error message. I have to click several time in the fields to make it disappear.
I attached a new screenrecord on the forge ticket.

capture du 2017-08-28 10-42-24

**Comment by Azouz-Jribi** [comment]
Hello @marionf 
Correction is done.

@eternoendless can you check it ?

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
Hello @Azouz-Jribi 
I still can save the carrier if the range are overlaping.
You should see the screenrecord11.mkv on the forge ticket.

**Comment by Azouz-Jribi** [comment]
Hello @marionf
Correction is done.

**Comment by marionf** [comment]
Thank you @Azouz-Jribi, I will re-test after the code review

**Comment by Azouz-Jribi** [comment]
Hi @eternoendless 
Rebase is done ;)

**Comment by eternoendless** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[js/admin/carrier_wizard.js, line 766 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8097#-KwFleDl62IvGMwcR9FP:-KwFleDl62IvGMwcR9FQ:bpsqz7m) ([raw file](https://github.com/prestashop/prestashop/blob/72e9eb92685bd875b7c55987d0417d4b8852fb6d/js/admin/carrier_wizard.js#L766)):*
> ```JavaScript
> var carriersRangeInputs = {
> 	/** Check the carriers range inputs after each change */
> 	'checkCarriersRangeInput': function() {
> ```

It's not necessary to add single quotes to object property names.

---

*[js/admin/carrier_wizard.js, line 766 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8097#-KwFlggycUkbdzmwbFYk:-KwFlggycUkbdzmwbFYl:b-3a396x) ([raw file](https://github.com/prestashop/prestashop/blob/72e9eb92685bd875b7c55987d0417d4b8852fb6d/js/admin/carrier_wizard.js#L766)):*
> ```JavaScript
> var carriersRangeInputs = {
> 	/** Check the carriers range inputs after each change */
> 	'checkCarriersRangeInput': function() {
> ```

I think the name of this method is misleading. The way it is now, it would make you believe it actually performs the validation, but it actually initializes the watch on inputs in order to validate them on change.

Method names should be an abstraction of what the method actually does.

---

*[js/admin/carrier_wizard.js, line 790 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8097#-Kw_epkzvWYijPGaiJ27:-Kw_epkzvWYijPGaiJ28:b-frgh1f) ([raw file](https://github.com/prestashop/prestashop/blob/72e9eb92685bd875b7c55987d0417d4b8852fb6d/js/admin/carrier_wizard.js#L790)):*
> ```JavaScript
> 			$(this).off();
> 			if ($('.wizard_error').length === 0) {
> 				index = $(this).closest('td').index();
> ```

Missing `var` before `index` here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8097)*

**Comment by Azouz-Jribi** [comment]
Hello @eternoendless 
Can you check it ?

Thanks ;)

**Comment by eternoendless** [comment]
Looks good to me

The code has grown too complex and needs refactoring though.

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8097#-:-Kwdq4mkwvMp3yox9bn4:b-u6jjo9)*

**Review by eternoendless — DISMISSED** [review]

**Comment by prasanthSelvar** [comment]
Hello @Azouz-Jribi,
I can still save carrier when I have the warning message "the range is not valid".
Especially in the case where the first value is greater than the second value.
See range.gif on the JIRA ticket [BOOM-3155](http://forge.prestashop.com/browse/BOOM-3155).

**Comment by Azouz-Jribi** [comment]
Hello @prasanthSelvar 
Correction is done ;)

**Comment by eternoendless** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8097)*

**Review by eternoendless — APPROVED** [review]

**Comment by eternoendless** [comment]
Thank you @Azouz-Jribi

**Comment by marzenarog** [comment]
> Questions	Answers
> Branch?	develop
> Description?	When you create or edit a carrier with invalid ranges, you should not be able to save it.
> Type?	bug fix
> Category?	BO
> BC breaks?	no
> Deprecations?	no
> Fixed ticket?	http://forge.prestashop.com/browse/BOOM-3155
> How to test?	Go to BO > shipping > carrier > Edit , add a new range and fill its input with incorrect values. Then, try to click another input field, the incorrect fields will be red, error will be displayed and the submit button will be disabled.
> This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8097)

Hello  guys!
I have same problem, could someone help me pls ( ranges are overlapping)  ?????   marzenarogus@gmail.com

**Pull request merged**